Canvas.toDataURL causes excessive GPUP memory use on Cocoa

nmoucht opened the pull request:
#### 55390aa216e406feeb3abdd5735de3da1130a8a2
<pre>
Canvas.toDataURL causes excessive GPUP memory use on Cocoa
<a href="https://bugs.webkit.org/show_bug.cgi?id=240639">https://bugs.webkit.org/show_bug.cgi?id=240639</a>
&lt;rdar://93567863&gt;

Unreviewed build fix after original patch was reverted.

From Kimmo&apos;s WIP: #1407

ImageBuffer::toData() and ImageBuffer::toDataURL() would construct a temporary
image-sized bitmap in the GPUP to act as the data encoding source image.
This intermediate bitmap is large for large source images, and will easily make GPUP
memory footprint exceed the limit.

Use the ImageBuffer drawing mechanism to composite the encoding source. This ensures
that the intermediate bitmaps are constructed similar to any other ImageBuffer, and as
such attributes the memory to WP.

Copy the ImageBuffer source to WP instead of GPUP and run the encoding in WP. This makes
the encoding cycles be spent in the correct process and reduces the code footprint of
GPUP.

ImageBuffer::toData() and ::toDataURL() were accessors that were propagated all the way
to the ImageBufferBackend and in GPUP case across processes. They are redundant, as
the data being used is available through copyNativeImage / copyImage.

* LayoutTests/fast/canvas/toDataURL-alpha-permutation-expected.html: Added.
* LayoutTests/fast/canvas/toDataURL-alpha-permutation.html: Added.
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* Source/WebCore/platform/graphics/ConcreteImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::copyImageBuffer):
(WebCore::copyNativeImageImpl):
(WebCore::ImageBuffer::create):
(WebCore::ImageBuffer::copyImage const):
(WebCore::ImageBuffer::toDataURL const):
(WebCore::ImageBuffer::toData const):
(WebCore::ImageBuffer::toDataURL):
(WebCore::ImageBuffer::toData):
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
* Source/WebCore/platform/graphics/cairo/ImageBufferCairoBackend.cpp:
(WebCore::ImageBufferCairoBackend::toDataURL const): Deleted.
(WebCore::ImageBufferCairoBackend::toData const): Deleted.
* Source/WebCore/platform/graphics/cairo/ImageBufferCairoBackend.h:
* Source/WebCore/platform/graphics/cairo/ImageBufferUtilitiesCairo.cpp:
(WebCore::imageTypeFromMIMEType):
(WebCore::platformMIMETypeIsJPEG):
(WebCore::encodeImage):
* Source/WebCore/platform/graphics/cairo/ImageBufferUtilitiesCairo.h:
* Source/WebCore/platform/graphics/cg/ImageBufferCGBackend.cpp:
(WebCore::createCroppedImageIfNecessary): Deleted.
(WebCore::colorSpaceForBitmap): Deleted.
(WebCore::ImageBufferCGBackend::copyCGImageForEncoding const): Deleted.
(WebCore::ImageBufferCGBackend::toData const): Deleted.
(WebCore::ImageBufferCGBackend::toDataURL const): Deleted.
* Source/WebCore/platform/graphics/cg/ImageBufferCGBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::invalidateCachedNativeImage const):
(WebCore::ImageBufferIOSurfaceBackend::putPixelBuffer):
(WebCore::ImageBufferIOSurfaceBackend::copyCGImageForEncoding const): Deleted.
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferUtilitiesCG.cpp:
(WebCore::jpegUTI):
(WebCore::utiFromImageBufferMIMEType):
(WebCore::platformMIMETypeIsJPEG):
(WebCore::encode):
(WebCore::encodeToVector):
(WebCore::encodeToDataURL):
(WebCore::data):
(WebCore::dataURL):
* Source/WebCore/platform/graphics/cg/ImageBufferUtilitiesCG.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::getDataURLForImageBuffer): Deleted.
(WebKit::RemoteRenderingBackend::getDataURLForImageBufferWithQualifiedIdentifier): Deleted.
(WebKit::RemoteRenderingBackend::getDataForImageBuffer): Deleted.
(WebKit::RemoteRenderingBackend::getDataForImageBufferWithQualifiedIdentifier): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/Shared/API/c/cg/WKImageCG.cpp:
(WKImageCreateDataURLFromImage):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::getDataURLForImageBuffer): Deleted.
(WebKit::RemoteRenderingBackendProxy::getDataForImageBuffer): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.cpp:
(WebKit::ImageBufferRemoteIOSurfaceBackend::toDataURL const): Deleted.
(WebKit::ImageBufferRemoteIOSurfaceBackend::toData const): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.h:

Canonical link: <a href="https://commits.webkit.org/252032@main">https://commits.webkit.org/252032@main</a>
</pre>

aperezdc commented on Source/WebCore/platform/MIMETypeRegistry.cpp:
```diff
@@ -832,4 +833,18 @@ Vector<String> MIMETypeRegistry::allowedFileExtensions(const Vector<String>& mim
+        return false;
+    return CFEqual(destinationUTI.get(), jpegUTI());
+#else
+    return mimeType.substring(sizeof "image") == "jpeg"_s;
```
It took me a while to realize that `sizeof "image"` will count _also_ the terminating nul character, and therefore the substring will be what follows after `"image/"`. Wouldn't it be better to actually check `mimeType == "image/jpeg"_s`? It's not like there are other MIME types for JPEG images.

nmoucht replied:
Fixed.

aperezdc commented on Source/WebCore/platform/MIMETypeRegistry.cpp:
```diff
@@ -832,4 +833,18 @@ Vector<String> MIMETypeRegistry::allowedFileExtensions(const Vector<String>& mim
+    return mimeType.substring(sizeof "image") == "jpeg"_s;
+#endif
+    ASSERT_NOT_REACHED();
+    return false;
```
These two lines are dead code, regardless of which part of the `#if` is built; they should be removed.

nmoucht replied:
Fixed.

nmoucht committed:
Canvas.toDataURL causes excessive GPUP memory use on Cocoa
https://bugs.webkit.org/show_bug.cgi?id=240639
<rdar://93567863>

Unreviewed build fix after original patch was reverted.

From Kimmo's WIP: #1407

ImageBuffer::toData() and ImageBuffer::toDataURL() would construct a temporary
image-sized bitmap in the GPUP to act as the data encoding source image.
This intermediate bitmap is large for large source images, and will easily make GPUP
memory footprint exceed the limit.

Use the ImageBuffer drawing mechanism to composite the encoding source. This ensures
that the intermediate bitmaps are constructed similar to any other ImageBuffer, and as
such attributes the memory to WP.

Copy the ImageBuffer source to WP instead of GPUP and run the encoding in WP. This makes
the encoding cycles be spent in the correct process and reduces the code footprint of
GPUP.

ImageBuffer::toData() and ::toDataURL() were accessors that were propagated all the way
to the ImageBufferBackend and in GPUP case across processes. They are redundant, as
the data being used is available through copyNativeImage / copyImage.

* LayoutTests/fast/canvas/toDataURL-alpha-permutation-expected.html: Added.
* LayoutTests/fast/canvas/toDataURL-alpha-permutation.html: Added.
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* Source/WebCore/platform/graphics/ConcreteImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::copyImageBuffer):
(WebCore::copyNativeImageImpl):
(WebCore::ImageBuffer::create):
(WebCore::ImageBuffer::copyImage const):
(WebCore::ImageBuffer::toDataURL const):
(WebCore::ImageBuffer::toData const):
(WebCore::ImageBuffer::toDataURL):
(WebCore::ImageBuffer::toData):
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
* Source/WebCore/platform/graphics/cairo/ImageBufferCairoBackend.cpp:
(WebCore::ImageBufferCairoBackend::toDataURL const): Deleted.
(WebCore::ImageBufferCairoBackend::toData const): Deleted.
* Source/WebCore/platform/graphics/cairo/ImageBufferCairoBackend.h:
* Source/WebCore/platform/graphics/cairo/ImageBufferUtilitiesCairo.cpp:
(WebCore::imageTypeFromMIMEType):
(WebCore::platformMIMETypeIsJPEG):
(WebCore::encodeImage):
* Source/WebCore/platform/graphics/cairo/ImageBufferUtilitiesCairo.h:
* Source/WebCore/platform/graphics/cg/ImageBufferCGBackend.cpp:
(WebCore::createCroppedImageIfNecessary): Deleted.
(WebCore::colorSpaceForBitmap): Deleted.
(WebCore::ImageBufferCGBackend::copyCGImageForEncoding const): Deleted.
(WebCore::ImageBufferCGBackend::toData const): Deleted.
(WebCore::ImageBufferCGBackend::toDataURL const): Deleted.
* Source/WebCore/platform/graphics/cg/ImageBufferCGBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::invalidateCachedNativeImage const):
(WebCore::ImageBufferIOSurfaceBackend::putPixelBuffer):
(WebCore::ImageBufferIOSurfaceBackend::copyCGImageForEncoding const): Deleted.
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferUtilitiesCG.cpp:
(WebCore::jpegUTI):
(WebCore::utiFromImageBufferMIMEType):
(WebCore::platformMIMETypeIsJPEG):
(WebCore::encode):
(WebCore::encodeToVector):
(WebCore::encodeToDataURL):
(WebCore::data):
(WebCore::dataURL):
* Source/WebCore/platform/graphics/cg/ImageBufferUtilitiesCG.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::getDataURLForImageBuffer): Deleted.
(WebKit::RemoteRenderingBackend::getDataURLForImageBufferWithQualifiedIdentifier): Deleted.
(WebKit::RemoteRenderingBackend::getDataForImageBuffer): Deleted.
(WebKit::RemoteRenderingBackend::getDataForImageBufferWithQualifiedIdentifier): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/Shared/API/c/cg/WKImageCG.cpp:
(WKImageCreateDataURLFromImage):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::getDataURLForImageBuffer): Deleted.
(WebKit::RemoteRenderingBackendProxy::getDataForImageBuffer): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.cpp:
(WebKit::ImageBufferRemoteIOSurfaceBackend::toDataURL const): Deleted.
(WebKit::ImageBufferRemoteIOSurfaceBackend::toData const): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.h:

Canonical link: https://commits.webkit.org/252032@main
